File: code/generate-sharp/utilities/operation-graph/OperationProxyInfo.cs
// <copyright file="OperationProxyInfo.cs" company="Soup">
// Copyright (c) Soup. All rights reserved.
// </copyright>

using System;
using System.Collections.Generic;
using System.Linq;

namespace Soup.Build.Utilities;

public record OperationProxyId(uint Value);

public class OperationProxyInfo : IEquatable<OperationProxyInfo>
{
	/// <summary>
	/// Initializes a new instance of the <see cref="OperationProxyInfo"/> class.
	/// </summary>
	public OperationProxyInfo() :
		this(
			new OperationProxyId(0),
			string.Empty,
			new CommandInfo(),
			[],
			new FileId(0),
			string.Empty,
			[],
			[])
	{
	}

	/// <summary>
	/// Initializes a new instance of the <see cref="OperationProxyInfo"/> class.
	/// </summary>
	public OperationProxyInfo(
		OperationProxyId id,
		string title,
		CommandInfo command,
		IList<FileId> declaredInput,
		FileId resultFile,
		string finalizerTask,
		ValueTable finalizerState,
		IList<FileId> readAccess)
	{
		this.Id = id;
		this.Title = title;
		this.Command = command;
		this.DeclaredInput = declaredInput;
		this.ResultFile = resultFile;
		this.FinalizerTask = finalizerTask;
		this.FinalizerState = finalizerState;
		this.ReadAccess = readAccess;
	}

	public bool Equals(OperationProxyInfo? other)
	{
		if (other is null)
			return false;

		var result = this.Id == other.Id &&
			this.Title == other.Title &&
			this.Command == other.Command &&
			Enumerable.SequenceEqual(this.DeclaredInput, other.DeclaredInput) &&
			this.ResultFile == other.ResultFile &&
			this.FinalizerTask == other.FinalizerTask &&
			this.FinalizerState == other.FinalizerState &&
			Enumerable.SequenceEqual(this.ReadAccess, other.ReadAccess);

		return result;
	}

	public override bool Equals(object? obj)
	{
		return Equals(obj as OperationProxyInfo);
	}

	public override int GetHashCode()
	{
		return (this.Id.GetHashCode() * 0x100000) +
			(this.Title.GetHashCode(StringComparison.InvariantCulture) * 0x1000) +
			this.Command.GetHashCode();
	}

	public static bool operator ==(OperationProxyInfo lhs, OperationProxyInfo other)
	{
		if (lhs is null)
			return other is null;
		return lhs.Equals(other);
	}

	public static bool operator !=(OperationProxyInfo lhs, OperationProxyInfo other)
	{
		return !(lhs == other);
	}

	public OperationProxyId Id { get; init; }
	public string Title { get; init; }
	public CommandInfo Command { get; init; }
	public IList<FileId> DeclaredInput { get; init; }
	public FileId ResultFile { get; set; }
	public string FinalizerTask { get; init; }
	public ValueTable FinalizerState { get; init; }
	public IList<FileId> ReadAccess { get; init; }
}

File: code/client/core/source/operation-graph/OperationProxyInfo.cpp
// <copyright file="OperationProxyInfo.cpp" company="Soup">
// Copyright (c) Soup. All rights reserved.
// </copyright>

module;

#include <chrono>
#include <map>
#include <string>

export module Soup.Core:OperationProxyInfo;

import Opal;
import :CommandInfo;
import :FileSystemState;
import :Value;

using namespace Opal;
using namespace std::chrono_literals;

export namespace Soup::Core
{
	using OperationProxyId = uint32_t;

	/// <summary>
	/// A node in the build graph that is the smallest operation proxy that will generate real operations
	/// </summary>
	class OperationProxyInfo
	{
	public:
		OperationProxyId Id;
		std::string Title;
		CommandInfo Command;
		std::vector<FileId> DeclaredInput;
		FileId ResultFile;
		std::string FinalizerTask;
		ValueTable FinalizerState;

		std::vector<FileId> ReadAccess;

	public:
		OperationProxyInfo() :
			Id(0),
			Title(),
			Command(),
			DeclaredInput(),
			ResultFile(),
			FinalizerTask(),
			FinalizerState(),
			ReadAccess()
		{
		}

		OperationProxyInfo(
			OperationProxyId id,
			std::string title,
			CommandInfo command,
			std::vector<FileId> declaredInput,
			FileId resultFile,
			std::string finalizerTask,
			ValueTable finalizerState,
			std::vector<FileId> readAccess) :
			Id(id),
			Title(std::move(title)),
			Command(std::move(command)),
			DeclaredInput(std::move(declaredInput)),
			ResultFile(resultFile),
			FinalizerTask(std::move(finalizerTask)),
			FinalizerState(std::move(finalizerState)),
			ReadAccess(std::move(readAccess))
		{
		}

		bool operator ==(const OperationProxyInfo& rhs) const
		{
			return Id == rhs.Id &&
				Title == rhs.Title &&
				Command == rhs.Command &&
				DeclaredInput == rhs.DeclaredInput &&
				ResultFile == rhs.ResultFile &&
				FinalizerTask == rhs.FinalizerTask &&
				FinalizerState == rhs.FinalizerState &&
				ReadAccess == rhs.ReadAccess;
		}
	};
}